### DISCUSSION TITLE
How to use the output of motion_correct on a different dataset?

### DISCUSSION THREAD

**Discussion by MartinKjunior** [dipy/dipy#3383] [Registration] [answered]
## Description
I have two sets of images of the same shape. One is the magnitude MRI data, second is phase MRI data. I would like to use the affine matrix that I got from running `motion_correct(nifti, gtab, ["center_of_mass", "translation", "rigid"])` on the magnitude data and apply it to the phase data. Is there a convenient method for doing that in DiPy? If not, could I suggest adding it in? 

I have not been able to find a general method for applying the affine matrix to a dataset. I am looking for a method that does not do any additional registration, just applies the already found affine. It seems to me like the solution might be along the lines of
```py
identity = np.eye(4)
affine_map = AffineMap(identity,
                       static.shape, static_grid2world,
                       moving.shape, moving_grid2world)
resampled = affine_map.transform(moving)
``` 
but if my final step was rigid, do I need to use `RigidTransform3D` or something else? I was able to do this in elastix using
```py
transformedImg = itk.transformix_filter(
            transformingImgitk,
            registeredTransformMap
            )
```
but I would like to move my code to DiPy.

**Comment by pjsjongsung** [comment]
Hi! I believe the code you mentioned above should work. (Of course except you want to change the identity affine to something you calculated from motion_correct) Have you tried and seen the output results?

**Reply by MartinKjunior** [reply]
Hi! Thanks for the quick reply. I have not tried it yet. I am unsure how to calculate the `..._grid2world` inputs. Could you help me with that?

**Reply by pjsjongsung** [reply]
With `motion_correct` you would have affine matrices that register each of the volumes to the mean b0. So right now I believe what you can do is for each volume, run your code while keeping `static_grid2world` an identity matrix and `moving_grid2world` the original affine from the file.

```
new_nifti, affines = motion_correct(nifti, gtab, ["center_of_mass", "translation", "rigid"])

phase_data, affine = load_nifti(phase_nifti)
resampled_phase_data = np.zeros_like(phase_data)
for i in range(phase_data.shape[-1]):
    affine_map = AffineMap(affines[i],
                           phase_data[..., i].shape, np.eye(4),
                           phase_data[..., i].shape, affines[i])
    resampled_phase_data[..., i] = affine_map.transform(phase_data[..., i])
```

Can you see if something like this works? @skoudoro might be able to provide more insight on this if this does not work as intended

**Reply by MartinKjunior** [reply]
Will give it a go tomorrow and let you know how it went. Thanks a lot!

**Reply by MartinKjunior** [reply]
> With `motion_correct` you would have affine matrices that register each of the volumes to the mean b0. So right now I believe what you can do is for each volume, run your code while keeping `static_grid2world` an identity matrix and `moving_grid2world` the original affine from the file.
> 
> ```
> new_nifti, affines = motion_correct(nifti, gtab, ["center_of_mass", "translation", "rigid"])
> 
> phase_data, affine = load_nifti(phase_nifti)
> resampled_phase_data = np.zeros_like(phase_data)
> for i in range(phase_data.shape[-1]):
>     affine_map = AffineMap(affines[i],
>                            phase_data[..., i].shape, np.eye(4),
>                            phase_data[..., i].shape, affines[i])
>     resampled_phase_data[..., i] = affine_map.transform(phase_data[..., i])
> ```
> 
> Can you see if something like this works? @skoudoro might be able to provide more insight on this if this does not work as intended

Tried this, the output is exactly the same as input. My code:

```py
import nibabel as nib
import numpy as np
from dipy.align.imaffine import AffineMap

reg_affine = np.load(R"...")
phase_data = nib.load(R"...").get_fdata()

resampled_phase_data = []
identity = np.eye(4)
shape = phase_data.shape[:-1]
for i in range(phase_data.shape[-1]):
    affine = reg_affine[..., i]
    affine_map = AffineMap(
        affine,
        shape, identity,
        shape, affine
    )
    resampled_phase_data.append(affine_map.transform(phase_data[..., i]))
resampled_phase_data = np.stack(resampled_phase_data, axis=3)
```

**Reply by MartinKjunior** [reply, marked as answer]
> With `motion_correct` you would have affine matrices that register each of the volumes to the mean b0. So right now I believe what you can do is for each volume, run your code while keeping `static_grid2world` an identity matrix and `moving_grid2world` the original affine from the file.
> 
> ```
> new_nifti, affines = motion_correct(nifti, gtab, ["center_of_mass", "translation", "rigid"])
> 
> phase_data, affine = load_nifti(phase_nifti)
> resampled_phase_data = np.zeros_like(phase_data)
> for i in range(phase_data.shape[-1]):
>     affine_map = AffineMap(affines[i],
>                            phase_data[..., i].shape, np.eye(4),
>                            phase_data[..., i].shape, affines[i])
>     resampled_phase_data[..., i] = affine_map.transform(phase_data[..., i])
> ```
> 
> Can you see if something like this works? @skoudoro might be able to provide more insight on this if this does not work as intended

I have figured it out now. The registration affine is in relation to the original image's location, so I have to use its affine as the domain_grid2world. Here is my solution (interpolation set to nearest because I need to avoid smoothing).
```py
import nibabel as nib
import numpy as np
from dipy.align.imaffine import AffineMap

reg_affine = np.load(R"...")
phase_nifti = nib.load(R"...")
mag_nifti = nib.load(R"...")

phase_data, phase_affine = phase_nifti.get_fdata(), phase_nifti.affine
mag_data, mag_affine = mag_nifti.get_fdata(), mag_nifti.affine

resampled_phase_data = []
identity = np.eye(4)
shape = phase_data.shape[:-1]
for i in range(phase_data.shape[-1]):
    affine = reg_affine[..., i]
    affine_map = AffineMap(
        affine,
        shape, mag_affine,
        shape, phase_affine
    )
    resampled_phase_data.append(
        affine_map.transform(
            phase_data[..., i],
            interpolation='nearest'
            )
        )
resampled_phase_data = np.stack(resampled_phase_data, axis=3)
```
